pkg/controller/master/node/maintain_controller.go

```diff
@@ -51,17 +61,87 @@ func (h *maintainNodeHandler) OnNodeChanged(_ string, node *corev1.Node) (*corev
+
+	// Wait until no VMs are running on that node.
+	vmiList, err := virtualmachineinstance.ListByNode(node, labels.NewSelector(), h.virtualMachineInstanceCache)
 	if err != nil {
 		return node, err
 	}
-	if len(vmis) != 0 {
+	if len(vmiList) != 0 {
 		return node, nil
 	}
+
+	// Restart those VMs that have been labeled to be shut down before
+	// maintenance mode and that should be restarted when the node has
+	// successfully switched into maintenance mode.
+	selector := labels.Set{util.LabelMaintainForceShutdownStrategy: "RestartOnEnable"}.AsSelector()
+	vmList, err := virtualmachine.ListByInstanceLabels(node.Namespace, selector, h.virtualMachineCache)
+	if err != nil {
+		return node, fmt.Errorf("failed to list VMs with labels %s: %w", selector.String(), err)
+	}
+	for _, vm := range vmList {
+		// Make sure that this VM was shut down as part of the maintenance
+		// mode of the given node.
+		if vm.Annotations[util.AnnotationMaintainForceShutdownNodeName] != node.Name {
+			continue
+		}
+
+		logrus.WithFields(logrus.Fields{
+			"namespace":           vm.Namespace,
+			"virtualmachine_name": vm.Name,
+		}).Info("restarting the VM that was temporary shut down for maintenance mode")
+
+		// Update the run strategy of the VM to start it and remove the
+		// annotation that was previously set when the node  went into
+		// maintenance mode.
+		// Get the running strategy that is stored in the annotation of the
+		// VM when it is shut down. Note, in general this is automatically
+		// patched by the VM mutator in general.
+		runStrategy := kubevirtv1.VirtualMachineRunStrategy(vm.Annotations[util.AnnotationRunStrategy])
+		if runStrategy == "" {
+			runStrategy = kubevirtv1.RunStrategyRerunOnFailure
+		}
+		vmCopy := vm.DeepCopy()
+		vmCopy.Spec.RunStrategy = &[]kubevirtv1.VirtualMachineRunStrategy{runStrategy}[0]
+		delete(vmCopy.Annotations, util.AnnotationMaintainForceShutdownNodeName)
+		_, err = h.virtualMachineClient.Update(vmCopy)
+		if err != nil {
+			return node, fmt.Errorf("failed to start VM %s/%s: %w", vm.Namespace, vm.Name, err)
+		}
+	}
+
 	toUpdate := node.DeepCopy()
 	toUpdate.Annotations[MaintainStatusAnnotationKey] = MaintainStatusComplete
 	return h.nodes.Update(toUpdate)
 }
+
+// OnNodeRemoved Ensure that all "harvesterhci.io/maintain-force-shutdown-node-name"
+// annotations on VMs are removed that are referencing this node.
+func (h *maintainNodeHandler) OnNodeRemoved(_ string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp == nil || node.Annotations == nil {
+		return node, nil
+	}
+
+	if _, ok := node.Annotations[MaintainStatusAnnotationKey]; !ok {
+		return node, nil
+	}
+
+	vms, err := h.virtualMachineCache.List(corev1.NamespaceAll, labels.Everything())
+	if err != nil {
+		return node, fmt.Errorf("failed to list VMs: %w", err)
+	}
+
+	for _, vm := range vms {
+		if vm.Annotations[util.AnnotationMaintainForceShutdownNodeName] != node.Name {
```

Comment:
> `AnnotationMaintainForceShutdownNodeNameis` implemented as an annotation because I do not have to search for this information and it is only metadata/additional data. In other words, exactly what annotations are intended for.

Actually, you do search for it here, right? If you set it on VM label and we can do this part like:

```go
	vms, err := h.virtualMachineCache.List(corev1.NamespaceAll, labels.Set{"harvesterhci.io/maintain-force-shutdown-node-name": node.Name}.AsSelector())
	if err != nil {
		return node, fmt.Errorf("failed to list VMs: %w", err)
	}

	for _, vm := range vms {
		vmCopy := vm.DeepCopy()
		delete(vmCopy.Annotations, util.AnnotationMaintainForceShutdownNodeName)
		_, err = h.virtualMachineClient.Update(vmCopy)
		if err != nil {
			return node, err
		}
	}
```